### PR TITLE
Add NeMo 2.0 on EKS example

### PR DESCRIPTION
**Pull request by olaoyea4**
*Issue #, if available:*

*Description of changes:*

This includes examples of how to run pretraining and finetuning jobs using NeMo 2.0 on SageMaker HyperPod EKS or a standalone EKS cluster

### CODE REVIEW

**Review by iankouls-aws — COMMENTED** [review]
I have tested this on two g5.8xlarge nodes as well as two p5e.48xlarge nodes. 

sample launch command (p5e.48xlarge):

```
    python pretrain_custom_dataset.py \
    --max_steps 200 \
    --nodes 2 \
    --gpus H200 \
    --gpu-devices 8 \
    --efa-devices 32 \
    --container_image $ACCOUNT_ID.dkr.ecr.$REGION.amazonaws.com/$REPO:$TAG \
    --env_vars_file env_vars.json \
    --pvc_name fsx-claim \
    --pvc_mount_path /mnt/nemo \
    --seq_length 1024 \
    --micro_batch_size 2 \
    --global_batch_size 512
```

Sample log output:

```
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default1]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7541:29615 [1] NCCL INFO Channel 09/0 : 0[1] -> 4[1] [send] via NET/AWS Libfabric/3/GDRDMA
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default1]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7541:29456 [1] NCCL INFO NET/OFI Global registrations supported
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default1]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7541:29615 [1] NCCL INFO Channel 13/0 : 0[1] -> 4[1] [send] via NET/AWS Libfabric/3/GDRDMA
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7540:29544 [0] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default6]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7546:29546 [6] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default2]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7542:29545 [2] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default4]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7544:29547 [4] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default3]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7543:29616 [3] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default7]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7547:29617 [7] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default5]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7545:29614 [5] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default1]:aws-nemo2-pretrain-20250625-22-26-324900e8-head:7541:29615 [1] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker ray-node [default2]:aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker:3580:25539 [2] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker ray-node [default6]:aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker:3584:25540 [6] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker ray-node [default0]:aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker:3578:25538 [0] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker ray-node [default4]:aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker:3582:25537 [4] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker ray-node [default3]:aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker:3581:25560 [3] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker ray-node [default5]:aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker:3583:25557 [5] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker ray-node [default1]:aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker:3579:25559 [1] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker ray-node [default7]:aws-nemo2-pretrain-20250625-22-26-324900e8-62d292-worker:3585:25558 [7] NCCL INFO Connected all trees
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:[NeMo W 2025-06-25 23:10:05 rerun_state_machine:1264] Implicit initialization of Rerun State Machine!
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:[NeMo W 2025-06-25 23:10:05 rerun_state_machine:239] RerunStateMachine initialized in mode RerunMode.DISABLED
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:Training epoch 0, iteration 0/199 | lr: 1.499e-07 | global_batch_size: 512 | global_step: 0 | reduced_train_loss: 11.02 | train_step_timing in s: 17.5
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:Training epoch 0, iteration 1/199 | lr: 2.999e-07 | global_batch_size: 512 | global_step: 1 | reduced_train_loss: 11.03 | train_step_timing in s: 2.558 | consumed_samples: 1024
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:Training epoch 0, iteration 2/199 | lr: 4.498e-07 | global_batch_size: 512 | global_step: 2 | reduced_train_loss: 11.03 | train_step_timing in s: 2.493 | consumed_samples: 1536
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:Training epoch 0, iteration 3/199 | lr: 5.997e-07 | global_batch_size: 512 | global_step: 3 | reduced_train_loss: 11.03 | train_step_timing in s: 2.522 | consumed_samples: 2048
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:Training epoch 0, iteration 4/199 | lr: 7.496e-07 | global_batch_size: 512 | global_step: 4 | reduced_train_loss: 11.02 | train_step_timing in s: 2.439 | consumed_samples: 2560
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:Training epoch 0, iteration 5/199 | lr: 8.996e-07 | global_batch_size: 512 | global_step: 5 | reduced_train_loss: 11.02 | train_step_timing in s: 2.421 | consumed_samples: 3072
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:Training epoch 0, iteration 6/199 | lr: 1.049e-06 | global_batch_size: 512 | global_step: 6 | reduced_train_loss: 11.01 | train_step_timing in s: 2.456 | consumed_samples: 3584
aws-nemo2-pretrain-20250625-22-26-324900e8-head ray-node [default0]:Training epoch 0, iteration 7/199 | lr: 1.199e-06 | global_batch_size: 512 | global_step: 7 | reduced_train_loss: 10.99 | train_step_timing in s: 2.464 | consumed_samples: 4096
```

**Review by iankouls-aws — APPROVED** [review]
The code is behaving as expected,
Tested both custom and mock dataset on g5 and p5 instances.
Merge approved.

Sample job status:

```
Experiment Status for aws-nemo2-pretrain-20250625-233459_1750894499

Task 0: training
- Status: RUNNING
- Executor: SkypilotExecutor
- Job id: aws-nemo2-pretrain-20250625-233459_1750894499___aws-nemo2-pretrain-20250625-233459_1750894499___training___1
```

File: 3.test_cases/megatron/nemo/kubernetes/README.md [hunk 1]
```diff
@@ -0,0 +1,694 @@
...
+  - For standard EKS: Follow [these instructions](https://github.com/aws-samples/awsome-distributed-training/tree/main/1.architectures/4.amazon-eks)
+    
+    **Additional Setup for Standard EKS (Non-HyperPod):**
+    
+    If you're using a standard EKS cluster (not SageMaker HyperPod), you'll need to install the following device plugins:
+    
+    1. **NVIDIA Device Plugin** (if not already installed):
+    ```bash
+    kubectl apply -f https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/v0.14.3/nvidia-device-plugin.yml
+    ```
```
Comment (KeitaW) [inline 3.test_cases/megatron/nemo/kubernetes/README.md:82]:
```suggestion
    1. **[NVIDIA device plugin](https://github.com/NVIDIA/k8s-device-plugin) and the NVIDIA GPU Operator** (if not already installed):
    ```bash
    # Add NVIDIA Helm repo
    helm repo add nvidia https://helm.ngc.nvidia.com/nvidia
    helm repo update
    # Install the GPU Operator in a namespace (e.g. gpu-operator)
    helm install gpu-operator nvidia/gpu-operator \
  --namespace gpu-operator --create-namespace \
  --wait
  ```

Better to use GPU operator per https://docs.skypilot.co/en/latest/reference/kubernetes/kubernetes-setup.html#kubernetes-setup-gpusupport.

File: 3.test_cases/megatron/nemo/kubernetes/README.md [hunk 2]
```diff
@@ -0,0 +1,694 @@
...
+```bash
+# Set environment variables
+ACCOUNT_ID=$(aws sts get-caller-identity --query Account --output text)
+REPO=aws-nemo
+TAG=24.12
+REGION=us-east-1  # Change to your desired region
+
+# Create ECR repository
+aws ecr create-repository --repository-name "$REPO" --region $REGION
+
```
Comment (KeitaW) [inline 3.test_cases/megatron/nemo/kubernetes/README.md:138]:
```suggestion
# Create ECR repository if it doesn't exist
aws ecr describe-repositories --repository-names ${REPO} --region ${REGION} 2>/dev/null || \
aws ecr create-repository --repository-name ${REPO} --region ${REGION}
```

File: 3.test_cases/megatron/nemo/kubernetes/README.md [hunk 3]
```diff
@@ -0,0 +1,694 @@
...
+> **Important**: Accelerator names must be **lowercase** (e.g., `l40s`, `h100`, `a100`) or SkyPilot will not recognize them.
+
+```bash
+# For L40S GPUs (g6e.x instances)
+kubectl label nodes hyperprod-i-0cebcc9ce37bb2fbf skypilot.co/accelerator=l40s --overwrite
+
+# For other GPU types (examples)
+# kubectl label nodes <node-name> skypilot.co/accelerator=h100 --overwrite
+# kubectl label nodes <node-name> skypilot.co/accelerator=a100 --overwrite
```
Comment (KeitaW) [inline 3.test_cases/megatron/nemo/kubernetes/README.md:205]: Manual deployment is a bit daunting. Should we use `sky.utils.kubernetes.gpu_labeler` in stead?
https://docs.skypilot.co/en/latest/reference/kubernetes/kubernetes-deployment.html#deploying-on-amazon-eks

File: 3.test_cases/megatron/nemo/kubernetes/README.md [hunk 4]
```diff
@@ -0,0 +1,694 @@
...
+3. **Data Formatting**: Converts dataset to JSONL format compatible with NeMo
+4. **Automated Processing Pod**: Deploys a Kubernetes pod for data processing
+
+### Automated Data Processing Script
+
+The `data-processing/data-processing.sh` script provides a complete automation solution for dataset preparation:
+
+```bash
+cd data-processing/
+chmod +x data-processing.sh
```
Comment (KeitaW) [inline 3.test_cases/megatron/nemo/kubernetes/README.md:262]:
```suggestion
```
Kindly commit the file after `chmod +x` instead.
